cluster/photon-controller/util.sh

```diff
@@ -0,0 +1,896 @@
+# Copyright 2016 The Kubernetes Authors All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# A library of helper functions that each provider hosting Kubernetes must implement to use cluster/kube-*.sh scripts.
+
+KUBE_ROOT=$(dirname "${BASH_SOURCE}")/../..
+source "${KUBE_ROOT}/cluster/photon-controller/config-common.sh"
+source "${KUBE_ROOT}/cluster/photon-controller/${KUBE_CONFIG_FILE-"config-default.sh"}"
+source "${KUBE_ROOT}/cluster/common.sh"
+
+PHOTON="photon -n"
+
+#####################################################################
+#
+# Public API
+#
+#####################################################################
+
+#
+# detect-master will query Photon Controller for the Kubernetes master.
+# It assumes that the VM name for the master is unique.
+# It will set KUBE_MASTER_ID to be the VM ID of the master
+# It will set KUBE_MASTER_IP to be the IP address of the master
+# If the silent parameter is passed, it will not print when the master
+# is found: this is used internally just to find the MASTER
+#
+function detect-master {
+    local silent=${1:-""}
+    KUBE_MASTER=${MASTER_NAME}
+    KUBE_MASTER_ID=${KUBE_MASTER_ID:-""}
+    KUBE_MASTER_IP=${KUBE_MASTER_IP:-""}
+
+    # We don't want silent failure: we check for failure
+    set +o pipefail
+    if [ -z ${KUBE_MASTER_ID} ]; then
+        KUBE_MASTER_ID=$($PHOTON vm list | grep "\tkubernetes-master\t" | awk '{print $1}')
+    fi
+    if [ -z ${KUBE_MASTER_ID} ]; then
+        kube::log::error "Could not find Kubernetes master node ID. Make sure you've launched a cluster with kube-up.sh"
+        exit 1
+    fi
+
+    if [ -z "${KUBE_MASTER_IP-}" ]; then
+        # Make sure to ignore lines where it's not attached to a portgroup
+        # Make sure to ignore lines that have a network interface but no address
+        KUBE_MASTER_IP=$($PHOTON vm networks ${KUBE_MASTER_ID} | grep -v "^-" | grep -E '\d+\.\d+\.\d+\.\d+' | head -1 | awk -Ft '{print $3}')
+    fi
+    if [ -z "${KUBE_MASTER_IP-}" ]; then
+        kube::log::error "Could not detect find Kubernetes master node IP. Make sure you've launched a cluster with 'kube-up.sh'" >&2
+        exit 1
+    fi
+    if [ -z $silent ]; then
+        kube::log::status "Master: $KUBE_MASTER ($KUBE_MASTER_IP)"
+    fi
+    # Reset default set in common.sh
+    set -o pipefail
+}
+
+#
+# detect-nodes will query Photon Controller for the Kubernetes minions
+# It assumes that the VM name for the minions are unique.
```

Comment:
Let's talk about these assumptions of unique VM names. I see them adding a lot of limitations to our usage of Kube-up in one Project